File: Part 1/code_from_book.cpp

// Chapter 1: Category

B f(A a);
C g(B b);

C g_after_f(A a) {
    return g(f(a));
}

template<class T> T id(T x) { return x; }

// Godbolt: https://godbolt.org/z/Yzf4Wh

std::reverse(v.begin(), v.end());
std::transform(v.begin(), v.end(), v.begin(), 
    [](auto e) { return e + 1; });

// Chapter 2: Types & Functions

// from book
int fact ( int n) {
    int result = 1;
    for (int i = 2; i <= n; ++ i)
        result *= i;
    return result;
}

// C++20: https://www.godbolt.org/z/xETPfo
using namespace std::ranges;

auto fact(int n) -> int {
    auto vals = views::iota(1, n);
    return std::reduce(begin(vals), end(vals), 1, std::multiplies{});
}

// C++23 potentially

using namespace std::ranges;

auto fact(int n) -> int {
    return views::iota(1, n) | fold(1, std::multiplies{});
}

// Chapter 9: Function Types

// https://www.godbolt.org/z/vj65KWdGx
auto sort_array_by_parity(auto& A) {
    std::ranges::partition(A, 
        [](auto e) { return e % 2 == 0; });   
    return A;
}

File: Part 1/code_from _book.hs

-- Chapter 1: Category

Prelude> :t (.)
(.) :: (b -> c) -> (a -> b) -> a -> c

-- Not from book
map (1+) . reverse

-- Chapter 2: Types & Functions

fact n = product [1 .. n]

-- Chapter 4: Kleisli Categories

type Writer a = (a, String )

( >=> ) :: (a -> Writer b) -> (b -> Writer c) -> (a -> Writer c)
m1 >=> m2 = \ x ->
    let (y, s1) = m1 x
        (z, s2) = m2 y
    in (z, s1 ++ s2)

return :: a -> Writer a
return x = (x, "")

upCase :: String -> Writer String
upCase s = (map toUpper s, "upCase ")

toWords :: String -> Writer [ String ]
toWords s = (words s, "toWords ")

process :: String -> Writer [ String ]
process = upCase >=> toWords
